Refactor ACM certificates and other changes

jamesmishra opened the pull request:
- Remove unnecessary Provose features. Remove hardcoded AWS AMIs
- actions: Upgrade Terraform version in fmt action
- Upgrade minimum Terraform version to 1.1.9
- Update Terraform providers and refactor deprecated code
- Run GItHub Actions on v4.0.x branch
- Do not create an ECR image repository policy when there are no principals
- Fix linting error
- Refactor ACM TLS certificates into a new top-level key

jamesmishra committed:
Refactor ACM TLS certificates into a new top-level key

We are now going to require users to manually specify the
TLS certificates that they want from AWS Certificate Manager (ACM).

Previously, we have been trying to automatically guess the
number of needed TLS certificates. This results in Terraform
frequently creating and destroying certificates, which ends
up running the user straight into AWS ACM's rate limits.

By manually specifying their certificate usage, users
can more carefully avoid the rate limits.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/provose/provose?pullRequest=13) <br/>All committers have signed the CLA.

jamesmishra committed:
iam: Fix Terraform drift caused by wrong sort order

It looks like Terraform expects JSON lists of strings
to be sorted.

jamesmishra committed:
containers: Resolve Terraform drift by specifying `volumesFrom` in our ECS container definition file

jamesmishra committed:
ecs: Resolve private image ECR tag into Docker image SHA-256 hashes

jamesmishra committed:
Remove old-school nested provider logic

We cannot use Terraform module `for_each` when we manually
set up Terraform providers across nested modules.

With this change, our submodules will automatically inherit
from the parent module.
